aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -145,13 +157,28 @@ module aptos_framework::object {
         from_bcs::to_address(hash::sha3_256(bytes))
     }
 
+    /// Derives an object address from type and source account: sha3_256([source | type | seed | 0xFC]).
+    public fun create_typed_object_address<T>(source: &address, seed: vector<u8>): address {
+        let bytes = bcs::to_bytes(source);
+        let type_name_bytes = bcs::to_bytes(&type_info::type_name<T>());
+        vector::append(&mut bytes, type_name_bytes);
+        vector::append(&mut bytes, seed);
+        vector::push_back(&mut bytes, OBJECT_FROM_TYPE_SCHEME);
+        from_bcs::to_address(hash::sha3_256(bytes))
+    }
+
```

Comment:
I think you want a witness...

`create_typed_object_address<T>(...., _witness: &T)`

alternatively... could we make an object witness such that I could do the following:

`create_deterministic_object(user: &signer, ref: &Witness): ConstructorRef`

where Witness is a ref that can be accessed either via `ConstructorRef or ExtendRef`...

Then an object implementation can expose a function like:
`fun generate_account_object(account: &signer): Object<T>`

which would create a fully useful object of say a fungible asset

Reply:
Seed is used to create a separate but deterministic wallet address for each fungible asset metadata object.

Reply:
Yep, an object implementation such as fungible assets might still want to create multiple objects with the same address formula (i.e. type T). In a way, type T allows namespacing objects by module/type.

Reply:
Switched over to using a DeriveRef to generate deterministic object addresses of this form:

hash(owner addr | address of object to derive from | 0xFC) where 0xFC is a new domain separator in object.move